### ISSUE TITLE
Tables in epubs aren't laid out properly

### ISSUE THREAD

**Issue by sam896** [sumatrapdfreader/sumatrapdf#1441]
A table in an epub 2 file wasn't laid out well.

The columns weren't lined up. All the text in a row was packed to the left - it wasn't lined up in columns at all. It was simple layout with a table tag, then tr tags containing td tags.

I've made a small example, but the file attach doesn't seem to be working in this new issue page - I'll try submitting it and see if I can then attach the file.

**Comment by sam896** [comment]
I can't seem to add the example file. I got a file selection dialogue by clicking the 'Attach file by...' button, but that didn't add the file, and dragging and dropping and pasting did nothing. Any suggestion how, please?

In the hopes it will be enough, as the rest of the epub I tried was a fairly plain epub version 2. the text for the table was:
    <?xml version="1.0" encoding="UTF-8" standalone="no"?><html xmlns="http://www.w3.org/1999/xhtml" xmlns:ops="http://www.idpf.org/2007/ops" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
    <head>
    <title>W</title>
    <meta content="application/xhtml+xml; charset=utf-8" http-equiv="Content-Type"/>
    </head>
    <body>
    <table border="0" cellspacing="0">
    <tr valign="top">
    <td align="center" style="border-bottom: 1px solid black;">A</td>
    <td align="center" style="border-bottom: 1px solid black;">B(<span class="small">2</span>)</td>
    <td align="center" style="border-bottom: 1px solid black;">C</td>
    </tr>
    <tr valign="top">
    <td align="left">2</td>
    <td align="left">2</td>
    <td align="left">2</td>
    </tr>
    <tr valign="top">
    <td align="left">2 22222222222 2222</td>
    <td align="left">2</td>
    <td align="left">2</td>
    </tr>
    <tr valign="top">
    <td align="left">2</td>
    <td align="left">2</td>
    <td align="left">2 22222222222 2222</td>
    </tr>
    <tr valign="top">
    <td align="left">ee</td>
    <td align="left">ee</td>
    <td align="left">e</td>
    </tr>
    </table>
    </body>
    </html>

**Comment by sam896** [comment]
I can't seem to add the example file. I got a file selection dialogue by clicking the 'Attach file by...' button, but that didn't add the file, and dragging and dropping and pasting did nothing. Any suggestion how, please?

In the hopes it will be enough, as the rest of the epub I tried was a fairly plain epub version 2. the text for the table was:
`
    <?xml version="1.0" encoding="UTF-8" standalone="no"?><html xmlns="http://www.w3.org/1999/xhtml" xmlns:ops="http://www.idpf.org/2007/ops" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
    <head>
    <title>W</title>
    <meta content="application/xhtml+xml; charset=utf-8" http-equiv="Content-Type"/>
    </head>
    <body>
    <table border="0" cellspacing="0">
    <tr valign="top">
    <td align="center" style="border-bottom: 1px solid black;">A</td>
    <td align="center" style="border-bottom: 1px solid black;">B(<span class="small">2</span>)</td>
    <td align="center" style="border-bottom: 1px solid black;">C</td>
    </tr>
    <tr valign="top">
    <td align="left">2</td>
    <td align="left">2</td>
    <td align="left">2</td>
    </tr>
    <tr valign="top">
    <td align="left">2 22222222222 2222</td>
    <td align="left">2</td>
    <td align="left">2</td>
    </tr>
    <tr valign="top">
    <td align="left">2</td>
    <td align="left">2</td>
    <td align="left">2 22222222222 2222</td>
    </tr>
    <tr valign="top">
    <td align="left">ee</td>
    <td align="left">ee</td>
    <td align="left">e</td>
    </tr>
    </table>
    </body>
    </html>
`

**Comment by GitHubRulesOK** [comment]
To supply a sample .epub etc it needs to be in a zip of less than 10Mb

**Comment by sam896** [comment]
Got that (~2KB zip containing the epub).

I click on the 'Attach files...' button just below this text box, a new window opens, I selet the file, click on Open, and the file window closes again with nothing added that I can see. There's a similar lack of success dragging an dropping: The icon of the file I'm dragging has a 'Move' tag added when I grad it over the button just below, but nothing's added when I drop.

Is the sample text in my second comment enough to work from? Presumably some of you already have or can create a skeleton epub to add it to. Or I could try something like base64 encoding and put it in as text!

**Comment by GitHubRulesOK** [comment]
Upload is probably blocked by some anti-viral setting as it most likely is running JavaScript to send file 

the table as given looks like this as supplied in its raw state which sounds like the description
![image](https://user-images.githubusercontent.com/12754913/73467952-f2193c80-437b-11ea-81df-1069311d13ea.png)

**Comment by sam896** [comment]
Yes, that is exactly the sort of issue I see with tables.

(I'll admit defeat on uploading files - I see you're good enough to cope with just the soure for that particular page.)

**Comment by kjk** [comment]
Currently Sumatra uses our own html renderer for epub which is not as good as mupdf's.

After 3.2 I'll likely switch to mupdf's epub renderer which should fix this issue.

**Comment by GitHubRulesOK** [comment]
A broader discussion on enhancing ePub support including tables such as these was covered in 
**This epub 3 file will not load and display correctly.** #849  so I am closing this as a duplicate of those requests

**Comment by selfmadebuba** [comment]
> Currently Sumatra uses our own html renderer for epub which is not as good as mupdf's.
> 
> After 3.2 I'll likely switch to mupdf's epub renderer which should fix this issue.

We have 3.4.6 and Tables are still rendered vertically instead of horizontally. What happened?

**Comment by selfmadebuba** [comment]
muPDF
![muPDF](https://github.com/sumatrapdfreader/sumatrapdf/assets/10959671/ff0742e8-6c7a-4f07-a708-e9bc51f7042a)

Sumatra 3.4.6
![sumatra 3 4 6](https://github.com/sumatrapdfreader/sumatrapdf/assets/10959671/b51c2ab4-9b4e-4727-a2e1-bf40bd004b85)

**Comment by selfmadebuba** [comment]
muPDF
![muPDF](https://github.com/sumatrapdfreader/sumatrapdf/assets/10959671/bde4c7dc-0a86-481a-abf8-0d68d3058fa5)


SumatraPDF 3.4.6
![sumatra 3 4 6](https://github.com/sumatrapdfreader/sumatrapdf/assets/10959671/bebf3e7e-550e-4ddc-9faf-9c548a00033e)

**Comment by selfmadebuba** [comment]
muPDF
![muPDF](https://github.com/sumatrapdfreader/sumatrapdf/assets/10959671/3e22fb93-9435-4901-a162-216c2b191edf)



SumatraPDF 3.4.6
![sumatra 3 4 6](https://github.com/sumatrapdfreader/sumatrapdf/assets/10959671/9122cc0f-0ebe-460f-bf1e-28e0d2755261)

**Comment by thinkfan** [comment]
@GitHubRulesOK @selfmadebuba   3.5 & 3.6 have already rendered epub-table horizontally, but it looks still ugly... (compared to [koodo](https://github.com/troyeguo/koodo-reader))
